### PR TITLE
Exchange AWS SAML from Okta for AWS Access Tokens via STS

### LINKED ISSUE #194: Implement getting SAML for AWS Auth from Okta IDP & role selection
Okta integrates with AWS to provide access to AWS over SAML (https://www.okta.com/integrations/aws-account-federation).

Unfortunately this workflow was not built with programmatic access in mind, so the current (and what previous iterations of BMX) do is some spooky HTML scraping for the SAML data. 

The folks at Okta do not consider this to be a workflow they have in mind either (https://devforum.okta.com/t/authenticating-a-saml-request-in-order-to-download-content-from-an-app/340/5), at some point we need to replace this with something more robust

### PR DESCRIPTION
**Pull request by boarnoah**
closes #194 

Add Cloud Provider (in this case AWS) which exchanges Okta provided SAML for short lived AWS tokens (these are what BMX outputs). 

- Parse out SAML from okta to Identify roles in the account,
- Use AWS SDK for .NET to Assume a role via the SAML

### CODE REVIEW

File: Bmx.CommandLine/CommandLine.cs [hunk 1]
```diff
@@ -14,7 +15,7 @@ public class CommandLine {
 
 		public CommandLine() {
 			// TODO: Fix this :| (ref: Program.cs)
-			_bmx = new BmxCore( new OktaClient( "d2l" ) );
+			_bmx = new BmxCore( new OktaClient( "d2l" ), new AwsClient() );
```
Comment (andrewalkema) [inline Bmx.CommandLine/CommandLine.cs:18]: I assume the `TODO` here is to inject these instead of using `new` and maybe move `"d2l"` to some sort of config?

Comment (boarnoah) [reply Bmx.CommandLine/CommandLine.cs:18]: It is a required parameter of bmx (from 2.0) so from launch args rather than config.

Constructor makes it mandatory to be set, my other alternative is to move setting Organization to some where after the OktaClient is initialized, it seems like a bit of a code smell since there is no way to enforce that organization is set prior to use of this class.

Comment (cfbao) [reply Bmx.CommandLine/CommandLine.cs:18]: > It is a required parameter of bmx (from 2.0) so from launch args rather than config.

Not sure I understand you correctly, but I'm pretty sure in bmx 2+ you can still set org in config.

Comment (boarnoah) [reply Bmx.CommandLine/CommandLine.cs:18]: ATM this rewrite doesn't have read from config yet.
When there isn't a config present its a required parameter with BMX 2.0. 

Regardless of how its set and read in it won't be possible to access organization to pass onto the constructor here so I would need to move the logic to set organization elsewhere.

Comment (MattPinelli) [reply Bmx.CommandLine/CommandLine.cs:18]: Also recall that this is an open source project - there shouldn't be any reference to `d2l` in the code.

Comment (cfbao) [reply Bmx.CommandLine/CommandLine.cs:18]: > Regardless of how its set and read in it won't be possible to access organization to pass onto the constructor here so I would need to move the logic to set organization elsewhere.

What I normally do is to read the config in the constructor. The config can be composed of different sources including config files and command line args.
More here: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1

Comment (boarnoah) [reply Bmx.CommandLine/CommandLine.cs:18]: Would like to handle proper loading of configs with #191 so different PR.

File: Bmx.Idp.Okta/OktaClient.cs [hunk 2]
```diff
@@ -75,10 +79,28 @@ public class OktaClient : IIdentityProvider {
...
+		private string ExtractAwsSaml( string htmlResponse ) {
+			// HTML page is fairly malformed, grab just the <input> with the SAML data for further processing
+			var inputRegexPattern = "<input name=\"SAMLResponse\" type=\"hidden\" value=\".*\"/>";
+			var inputRegex = new Regex( inputRegexPattern, RegexOptions.Compiled );
+
+			// Access the SAML data from the parsed <input>
+			var inputXml = new XmlDocument();
+			inputXml.LoadXml( inputRegex.Match( htmlResponse ).Value );
+
+			return inputXml.SelectSingleNode( "//@value" ).InnerText;
```
Comment (neverendingqs) [inline Bmx.Idp.Okta/OktaClient.cs:101]: Thoughts on parsing the SAML response earlier, or adjusting the naming? It's not clear beyond this point that we're passing around SAML response statements until we parse it in `SetSamlToken()`.

File: Bmx.Service.Aws/AwsClient.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+				_awsRoles.Add( new AwsRole {
+					PrincipalArn = nodeContents[0],
+					RoleArn = nodeContents[1],
+					RoleName = nodeContents[1].Split( "/" )[1]
+				} );
+			}
+
+			return _awsRoles.Select( role => role.RoleName ).ToArray();
+		}
```
Comment (neverendingqs) [inline Bmx.Service.Aws/AwsClient.cs:59]: The way `_samlToken` and `_awsRoles` is set suggests that we should use a builder pattern maybe? Something something

```cs
class AwsClient {
  public AwsClient ( AwsRoles[] awsRoles ) {
    _awsRoles = awsRoles
    _stsClient = new AmazonSecurityTokenServiceClient();
  }
  ...
  static AwsClient Create( string encodedSaml ) {
    ...
    var samlToken = LoadXml( ... );
    ...
    var awsRoles = ....
    ...
    return new AwsClient( awsRoles );
  }
}
```
?

This will prevent callers from calling `GetRoles()` or `GetTokens()` before the internals are fully set up.

File: Bmx.Idp.Okta/OktaClient.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (boarnoah) [reply Bmx.Idp.Okta/OktaClient.cs:101]: Yes, its not ideal but as far as I can tell they decided to do so as there was no viable alternatives. 

Mentioned in the ticket #194 there might be a different way we could pursue at a later point (unsure as to the scope of work to do so).

File: Bmx.Service.Aws/AwsClient.cs (hunk 3 above)
Comment (boarnoah) [reply Bmx.Service.Aws/AwsClient.cs:59]: Thanks! I really didn't like how I had it setup, this pattern solves that problem very well.

File: Bmx.Idp.Okta/OktaClient.cs (hunk 2 above)
Comment (boarnoah) [reply Bmx.Idp.Okta/OktaClient.cs:101]: I should clarify the variable names here, Identity provider at this point is dealing with HTML (inputXML) is an HTML<input> node. 

I used XPath parsing  at this point since it was a straightforward way to get the contents of the <input> node's Value attribute.

As to why Identity Provider is passing the token out as a encoded string, initially I had it setup where it did pass parsed SAML as an XML doc, but it requires writing a lot of additional things to re-encode and export the XMLDocument back to a string that the AWS API accepts.

File: Bmx.Service.Aws/AwsClient.cs (hunk 3 above)
Comment (boarnoah) [reply Bmx.Service.Aws/AwsClient.cs:59]: While trying to implement DI and Unit Tests I've realized this issue extends deeper than just this setting. 

Services for Okta and AWS both are stateful, and there is a very strong temporal coupling (methods needing to be called in a very specific order for the app to work without runtime errors). 

As far as I can see there is no way to remove this coupling without making the services stateless (and by leaking internal implementation details from those classes into BmxCore).

tl;dr its an extensive issue that goes deeper than just having the presence encodedSaml guaranteed via builder pattern, requires investigation on its own IMO.

Comment (neverendingqs) [reply Bmx.Service.Aws/AwsClient.cs:59]: Oof gotcha. Makes sense to me.
